### PR TITLE
Journal: Website Url for journal link posts not being saved to ItemData in Journal database - Fixes #3308 

### LINKED ISSUE #3308: links posted in Journal are not clickable and the link created goes to a page on the site
## Description of bug
Posting a link to a site in the journal posts content about the site, however the link to the site is not correct when you click on it, and the link in journal does not work.

## Steps to reproduce
List the steps to reproduce the behavior:
1. Go to journal
2. post a link
3. click on link
4. takes you to a page on your site 

## Current result
Link does not work properly that is created, and the link posted is not clickable

## Expected result
A possibly shortened clickable link with information about the page and a link to the page posted in the header

## Screenshots
If applicable, provide screenshots to help explain your problem.
image

### PR DESCRIPTION
**Pull request by thabaum**
## Summary

Fixes #3308 

`url` string was being checked if it had `//` so any address with it would get denied.  

I changed the statement that did NOT allow urls with `//` to allow  `HTTPS://`, `HTTP://` and `www.`

image

Sure seems to show up the links now in the database and when testing.

I am seeing the effects of my changes after rebuilding the project.  This is also my first attempt in long time to do a PR.  I found it easiest since only 1 file to copy and paste the changed file contents until I get used to github branches again.

I changed the not allow to allow urls with `HTTPS://`, `HTTP://` and `www.`

I will try to make these also create a link for the address typed in the message.  I hope this is acceptable.  I also changed a few names to have UppercaseFormat and was going to document what I could but I need to learn how AND I wanted to do some dependency injection fixes on that same file.

If I do before this gets applied, if it does, I will resubmit a new PR with all the changes made.

### CODE REVIEW

**Review by david-poindexter — CHANGES_REQUESTED** [review]
Thank you for the PR @thabaum - I made a few suggestions.  Ultimately, it seems we need a better method to better determine what an "allowed link" actually is, but for now, I would recommend we stick with a simple fix.

File: DNN Platform/Modules/Journal/ServicesController.cs
Comment (valadas) [inline DNN Platform/Modules/Journal/ServicesController.cs:321]: I would do this instead to support all different types of URLs (except file, which would be local to the client).

```suggestion
            Uri uriResult;
            return Uri.TryCreate(url, UriKind.Absolute, out uriResult)
                && uriResult.Scheme != Uri.UriSchemeFile;
```

File: DNN Platform/Modules/Journal/ServicesController.cs
Comment (valadas) [inline DNN Platform/Modules/Journal/ServicesController.cs]: Can we remove this whitespace
```suggestion
```

**Review by valadas — CHANGES_REQUESTED** [review]
I think we can improve the validity of URLs (see comment on that actual method)

Also, I was worried a little bit about XSS but verified it is handled higher up the stack, so other than this, the PR looks fine to me.

File: DNN Platform/Modules/Journal/ServicesController.cs
Comment (david-poindexter) [reply DNN Platform/Modules/Journal/ServicesController.cs:321]: I will resolve this one in favor of recommendation by @valadas

Comment (valadas) [reply DNN Platform/Modules/Journal/ServicesController.cs:321]: I _think_ I have a better solution in my suggestion, with my suggested change it would support all non-local URL schemes like:
- http
- https
- ftp
- news
- mailto

Comment (david-poindexter) [reply DNN Platform/Modules/Journal/ServicesController.cs:321]: Great suggestion @valadas 🎉 🌮

Comment (thabaum) [reply DNN Platform/Modules/Journal/ServicesController.cs:321]: you are leaving out the www.

Comment (thabaum) [reply DNN Platform/Modules/Journal/ServicesController.cs:321]: i tried this... didnt work for the www for me so I created what I made.  I will test out @valadas solution.

Comment (valadas) [reply DNN Platform/Modules/Journal/ServicesController.cs]: If other reviewers have concerns with allowing too many schemes, then this can also be rewritten to allow just what we want instead of not allowing what we want (reverse the logic)

Comment (thabaum) [reply DNN Platform/Modules/Journal/ServicesController.cs]: I am all for improving it, I figured someone was going to toss us the one up here if nothing else let this fly possibly as is.  I am going to test this solution now.  Thanks for all the quick responses.  I would love to put this one to bed!

File: DNN Platform/Modules/Journal/ServicesController.cs
Comment (valadas) [inline DNN Platform/Modules/Journal/ServicesController.cs:322]: Actually re-thinking about this, it is best practice to use allow-lists instead of deny-lists from a security perspective...

```suggestion
            return Uri.TryCreate(url, UriKind.Absolute, out uriResult)
                && (
                    uriResult.Scheme == Uri.UriSchemeHttp ||
                    uriResult.Scheme == Uri.UriSchemeHttps ||
                    uriResult.Scheme == Uri.UriSchemeFtp ||
                    uriResult.Scheme == Uri.UriSchemeMailto ||
                    uriResult.Scheme == Uri.UriSchemeNews ||
                    uriResult.Scheme == Uri.UriSchemeNntp
                );
```

**Review by valadas — CHANGES_REQUESTED** [review]
Change to use an allow-list instead of a deny-list

File: DNN Platform/Modules/Journal/ServicesController.cs
Comment (david-poindexter) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: This seems more explicit.  Thanks @valadas 🎉 🌮

Comment (david-poindexter) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: CC: @mitchelsellers @bdukes

Comment (thabaum) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: this is what I was thinking of building from scratch, gave up and made the simple fix I put here.  I was hoping some other community ideas would make this right.

Comment (thabaum) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: The other solution works and the uri namespace is what I need to get familiar with.  I have a few more tweaks I want to do using pretty much this logic.

Comment (thabaum) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: I tested the first solution ok.  Testing your new one now.

Comment (thabaum) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: Seems to work good, thank you!

Comment (thabaum) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: if everyone agrees and nothing better is on its way anytime soon I will commit suggestion.

Comment (david-poindexter) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: @thabaum we chatted and would like to have @mitchelsellers and/or @bdukes review before committing this suggestion.  Thanks.

Comment (thabaum) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: as far as allow list

> Change to use an allow-list instead of a deny-list

I believe this is pretty much what I did.

Comment (bdukes) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: Do we want to support all of those protocols? I would think `http`, `https`, and `mailto` would be all we'd need

Comment (valadas) [reply DNN Platform/Modules/Journal/ServicesController.cs:322]: I don't really mind, just went for everything in that namespace that was not something local... I could see an FTP link to download something off a public FTP for instance...

**Comment by valadas** [comment]
I put this pr on hold, we need to run the integration tests manually for the journal before merging this...

**Comment by thabaum** [comment]
Please include fix #4960 when you merge this one, and test together.  Thank you.

**Comment by mitchelsellers** [comment]
Upon a discussion in the approver's meeting, there was reasoning behind not linking these by default.  

We believe the proper solution for this would be to add a setting to enable or disable this on the site level.  Ideally off by default and only settable by a host user.

**Comment by thabaum** [comment]
I was thinking the same thing honestly, it would be nice to have the option.   Also the option to not have links displayed period with images.

So where does the on/off switch go?  

In other words when journal module `Activate Links` option is off:
Do we want to let it do what it is currently doing with the `link` template being parsed when there is a link?

 Or not show the `link` template when option is off?

**Comment by mitchelsellers** [comment]
The ideal location for this, I believe, is under the "Security" settings for the portal with all other security related configuration elements in the PersonaBar

**Comment by thabaum** [comment]
You may be right since it works with profile activity walls.

**Comment by thabaum** [comment]
Should we turn off the template where it pulls an image, title and description from a website when a link is typed.

This is entered into the database via JSON `itemData` we are looking to create a second `Url` and `Title` for a backlink to a social group in the issue below:

https://github.com/dnnsoftware/Dnn.Platform/issues/3035

Any feedback here is appreciated.  Thank you.

**Comment by dustinp26** [comment]
> 

I would agree. Right now the sharing of a comment in a group provides no context when shared to the main journal feed. 

Another item that would be nice is ability to add reply's to comments where it can auto tag the uses comment you are replying to with an @user as well as ability to have option for comments to include photos, attachments, etc.

**Comment by thabaum** [comment]
> Another item that would be nice is ability to add reply's to comments where it can auto tag the uses comment you are replying to with an @user as well as ability to have option for comments to include photos, attachments, etc.

I am not sure if there is an issue created for the @mentions but this is an area of concern for me as well going forward.  Issue I see is when you go switch from username to email address for logging into a DNN system it erases the username for any @mentions to work.  So I think maybe a field would need selected in a profile registration form if anyone even wants to be "mentioned" and make it something that is called "handle" like a trucker uses while on the radio.

I feel a number of PR's coming as soon as I crack this all open again.  I think that is my solution is to use a selected profile field as the place a user can enter a "handle" and can leave blank if they just want to be left alone.

____________________________

Back to this PR what is this switch going to be named default?

`Activate Links in Journal`

Under 

`Enable Remember Me on Login Control`

image

Or we talking a whole new journal tab. 

Last is does this turn of the collection of the Url like it currently does OR do we want this to shut off the use of the link template and use the basic one which would not utilize the backlinks.

@mitchelsellers 
Cant you just turn it off by editing the resx file entries just by removing all the links and images and put the default comment template?  That pretty much dictates what is displayed if I am not mistaken.  I have not tried it but I imagine if you copy one to the other there won't be any link `ItemData` displayed.

hmmm...

**Comment by thabaum** [comment]
so another approach is having the switch just use the basic `journal_post` language file instead of the link?  

Should it also turn off the collection of the link?

I can put together some documentation how to edit the way the journal creates posts and the administrator can handle it how they like.  If they change their mind the links can still work that way.  

It just seems like the language files should handle what is shown or not shown.  A user cannot change this.

image

**Comment by thabaum** [comment]
@valadas #4960 

I have all links converting to start with `//` in the `href` and `img src` areas of the journal when creating links.

`return !string.IsNullOrEmpty(url) && url.Contains("//")`

This works, however
`
            Uri uriResult;
            return Uri.TryCreate(url, UriKind.Absolute, out uriResult)
                    && (
                        uriResult.Scheme == Uri.UriSchemeHttp ||
                        uriResult.Scheme == Uri.UriSchemeHttps ||
                        uriResult.Scheme == Uri.UriSchemeFtp ||
                        uriResult.Scheme == Uri.UriSchemeMailto ||
                        uriResult.Scheme == Uri.UriSchemeNews ||
                        uriResult.Scheme == Uri.UriSchemeNntp
                        );
`

Does not.

Since changes were needed to the other PR to change links to start with `//` 

`
            return Uri.TryCreate(url, UriKind.Absolute, out uriResult)
                    && ((
                        uriResult.Scheme == Uri.UriSchemeHttp ||
                        uriResult.Scheme == Uri.UriSchemeHttps ||
                        uriResult.Scheme == Uri.UriSchemeFtp ||
                        uriResult.Scheme == Uri.UriSchemeMailto ||
                        uriResult.Scheme == Uri.UriSchemeNews ||
                        uriResult.Scheme == Uri.UriSchemeNntp)
                        || url.StartsWith("//"));
`
would allow these links and keeps the journal link functioning.

@mitchelsellers 
I am still not too clear if a tutorial in DNN Docs would suffice for showing how to disable link functionality in the template language resource files if it is a security concern.  But I would be willing to create something that would work and submit it if this PR is approved.

The feature for security on/off I will to add later is for functionality that is auto creating links in the posts and comments for all links, but not for site descriptions and images.  I can see this being added to the switch regardless.  I am just not sure that should be done to link info collected, you still need to change the template if you dont want it to show all this stuff. 

Unless we make it so it wont collect the `ItemData`?  Then that would be another switch in the personaBar settings.  I would need to get to work on that part of the PR if this is truely what is desired and simply changing the template file is not enough...

I see two switches if we want both as I do plan to add the linking in journal text feature if allowable.

**Comment by thabaum** [comment]
> We believe the proper solution for this would be to add a setting to enable or disable this on the site level. Ideally off by default and only settable by a host user.

I read into this one liner a bit more... before I splurge some time into this unless I hear otherwise I will create an option in the persona bar as follows:

1. Only the host can change the option
2. Collecting link data is off by default (will treat the post as a normal journal_post and not look for images)
3. Option located in "Security" settings.

**Comment by thabaum** [comment]
I have been too busy to look at this however I should be able to start working on it again to wrap this up hoping in a week or two if not sooner.  Simply had way too much on my plate last couple months.  Just a heads up.  It is still on my to do list, Thank you.

**Comment by thabaum** [comment]
Thank you for not deleting this PR yet, I apologize for not being available to contribute over the past 6 months due to a number of unforeseen events out of my control.  This is on the back burner as of now but I promise a lot of great contributions soon as things normalize allowing me the time to donate and make this fix along with others.

HAPPY 4th to the GREATEST country on the planet!  GO USA!

**Comment by mitchelsellers** [comment]
I am going to close this PR for now.  @thabaum if you get around to having time for this one, please don't hesitate to re-open and re-submit.

**Comment by thabaum** [comment]
I plan on attacking this again by end of next week, I need to study where I left off. Sorry for the delays I will shoot for next release date for some love here as I am just now able to clear up some of my schedule for this I am hoping.

**Pull request closed without merging**
